### ISSUE TITLE
CSI plugin supervisor can spawn one extra goroutine

### ISSUE THREAD

**Issue by zizon** [hashicorp/nomad#12772]
### Nomad version
Output from `nomad version`

### Issue
As state in #12744 the corresponding fix #12752 simply keep `csiPluginSupervisorHook` in state un-done. 

This ensure `csiPluginSupervisorHook` prestart runs each time when the task runner stay in [main loop](https://github.com/hashicorp/nomad/blob/e4d6d51035629be585294ff7fc63885d7bca9824/client/allocrunner/taskrunner/task_runner.go#L544).

But aslo the `ensureSupervisorLoop` gorouting of `csiPluginSupervisorHook` spawn in [poststart](https://github.com/hashicorp/nomad/blob/e4d6d51035629be585294ff7fc63885d7bca9824/client/allocrunner/taskrunner/task_runner.go#L579). 

The ensureSupervisorLoop once run into [tight loop](https://github.com/hashicorp/nomad/blob/main/client/allocrunner/taskrunner/plugin_supervisor_hook.go#L321), the only chance to terminating such goroutine is by canceling the [shutdownCtx](https://github.com/hashicorp/nomad/blob/main/client/allocrunner/taskrunner/plugin_supervisor_hook.go#L236). 

Such context is shutdown only in the [Stop](https://github.com/hashicorp/nomad/blob/main/client/allocrunner/taskrunner/plugin_supervisor_hook.go#L447) method of `csiPluginSupervisorHook`, which in turn as far as I know got called after the [task runner exhaust the CSI retries](https://github.com/hashicorp/nomad/blob/e4d6d51035629be585294ff7fc63885d7bca9824/client/allocrunner/taskrunner/task_runner.go#L663).

**Comment by tgross** [comment]
Hi @zizon! Thanks for opening this issue with some more detail.

> The ensureSupervisorLoop once run into tight loop, the only chance to terminating such goroutine is by canceling the shutdownCtx.
>
> Such context is shutdown only in the Stop method of csiPluginSupervisorHook, which in turn as far as I know got called after the task runner exhaust the CSI retries.

You're absolutely right there, but I think what you've missed is that we don't create a _new_ supervisor loop if we're already running one because of the check at the start of [`Poststop`](https://github.com/hashicorp/nomad/blob/e4d6d51035629be585294ff7fc63885d7bca9824/client/allocrunner/taskrunner/plugin_supervisor_hook.go#L225-L232). We just reuse the one we had before.

I wanted to verify this for myself, so I deployed a plugin and then brought up all the goroutines via the [pprof endpoint](http://localhost:4646/debug/pprof/goroutine?debug=2). I can find the supervisor loop goroutine here:

```
goroutine 1054 [select]:
github.com/hashicorp/nomad/client/allocrunner/taskrunner.(*csiPluginSupervisorHook).ensureSupervisorLoop(0xc0006b9080, {0x2b02cd8, 0xc000931a40})
	github.com/hashicorp/nomad/client/allocrunner/taskrunner/plugin_supervisor_hook.go:320 +0x8d6
created by github.com/hashicorp/nomad/client/allocrunner/taskrunner.(*csiPluginSupervisorHook).Poststart
	github.com/hashicorp/nomad/client/allocrunner/taskrunner/plugin_supervisor_hook.go:234 +0x128
```

Then I killed the task via `docker kill` and waited for it to restart, and checked the goroutine trace again. We don't start a new plugin supervisor goroutine because although the _container_ has failed the task runner is still going. So we still have the exact same goroutine (1054) running and I can see that it's still being used to send fingerprinting RPCs to the running plugin.

If I kill the container a few more times, I eventually exceed the restart limit:

```
Recent Events:
Time                       Type                   Description
2022-04-25T11:10:26-04:00  Not Restarting         Exceeded allowed attempts 2 in interval 30m0s and mode is "fail"
2022-04-25T11:10:26-04:00  Terminated             Exit Code: 137, Exit Message: "Docker container exited with non-zero exit code: 137"
2022-04-25T11:09:59-04:00  Started                Task started by client
2022-04-25T11:09:58-04:00  Restarting             Task restarting in 18.59534704s
2022-04-25T11:09:58-04:00  Terminated             Exit Code: 137, Exit Message: "Docker container exited with non-zero exit code: 137"
2022-04-25T11:05:24-04:00  Started                Task started by client
2022-04-25T11:05:24-04:00  Restarting             Task restarting in 17.235255963s
2022-04-25T11:05:24-04:00  Terminated             Exit Code: 137, Exit Message: "Docker container exited with non-zero exit code: 137"
2022-04-25T11:04:52-04:00  Plugin became healthy  plugin: org.democratic-csi.nfs
2022-04-25T11:04:42-04:00  Started                Task started by client
```

And at that point I get a new plugin along with a new supervisor loop goroutine, but the other one has exited as expected.

**Comment by zizon** [comment]
I see, the `running` stay false until the early stage of `ensureSupervisorLoop`.

A rare case is that,
1. go ensureSupervisorLoop() was called, but not yet scheduled by go runtime
2. after returning from poststop, the task fails immediately.
3. if after that monument a retry occurs, yet another ensureSupervisorLoop will be spawned.
4. when this retry restart succeed, there will be possible two ensureSupervisorLoop got running, both got schedule by runtime.

But I don`t think it will happens in real life.

**Comment by tgross** [comment]
That's true, you could get slightly overlapping goroutines in that case. If the original goroutine wins the race for the lock, it would then exit because its stop context has been closed. But if the new goroutine wins the race for the lock, the original goroutine would end up blocking forever waiting on that lock until the task ends. And at that point it would fall thru because of the stop context.

This is a very narrow and unlikely race, with limited scope; we can leak at most one goroutine per plugin task. But it is a bug so I'm going to reopen the issue (and retitle it for clarity). Just a warning that the fix probably won't land in Nomad 1.3.0 GA just because of our schedule, but I'll try to get it in.

**Comment by zizon** [comment]
By the way, my original thoughts about fixing #12744 is to find a way to persist [hookResources](https://github.com/hashicorp/nomad/blob/e4d6d51035629be585294ff7fc63885d7bca9824/client/allocrunner/taskrunner/task_runner.go#L167) in state, which expect to help drivers recover appropriate configs(in that case, the mounts csi hook added).

But it would break backward compatible since the serialization scheme changes.
